daceml/autodiff/library/python_frontend.py

```diff
@@ -0,0 +1,198 @@
+                       sdfg.make_array_memlet(grad_name))
+
+
+@op_repository.replaces_attribute('Array', 'grad')
+def grad(pv: 'ProgramVisitor', sdfg: SDFG, state: SDFGState, arr: str) -> str:
+    """
+    Returns the name of the gradient buffer of the given array.
+
+    The Array must have been marked as requires_grad_ using
+    ``arr.requires_grad_()``, otherwise there will be an error
+    """
+
+    if arr not in sdfg.arrays:
+        raise common.DaceSyntaxError(pv, None,
+                                     "Array {} is not defined".format(arr))
+    desc = sdfg.arrays[arr]
+    if not isinstance(desc, Array):
+        raise common.DaceSyntaxError(
+            pv, None,
+            "Called .grad on an Array that was not a Parameter. Convert it to a parameter "
+            " first using .requires_grad_()")
+
+    return desc.gradient
+
+
+@op_repository.replaces_method('Array', 'requires_grad_')
+@op_repository.replaces_method('Scalar', 'requires_grad_')
+def requires_grad_(pv: newast.ProgramVisitor, sdfg: SDFG, state: SDFGState,
+                   self: str):
+    """
+    Converts a array to a Parameter (i.e.
+    :class:``daceml.autograd.library.Array``). This creates a descriptor for
+    the gradient buffer for this array.
+    """
+
+    if self not in sdfg.arrays:
+        raise common.DaceSyntaxError(pv, None,
+                                     "Array {} is not defined".format(self))
+    Array.make_parameter(sdfg, self)
+
+
+@op_repository.replaces_method('Array', 'backward')
+@op_repository.replaces_method('Scalar', 'backward')
+def backward_method(pv: newast.ProgramVisitor,
+                    sdfg: SDFG,
+                    state: SDFGState,
+                    self: str,
+                    grad: Optional[str] = None):
+    """
+    Alias for ``torch.autograd.backward(self)``
+    """
+    backward(pv, sdfg, state, self, grad)
+
+
+# FIXME hack to make sure BackwardPass nodes are expanded before other library nodes when using the python frontend
+_current_opt = config.Config.get("optimizer", "interface")
+if _current_opt is not None:
+    warnings.warn(
+        "DaCeML autodiff doesn't work with dace optimizers at the moment, optimizer '{}' will be disabled"
+        .format(_current_opt))
+
+config.Config.set(
+    "optimizer",
+    "interface",
+    value="daceml.autodiff.library.python_frontend.BackwardPassExpander")
+config.Config.set("optimizer", "transform_on_call", value=True)
+
+
+class BackwardPassExpander(optimizer.Optimizer):
+    """
+    An optimizer that expands the backward pass nodes of an SDFG.
+
+    Since optimizers are called before other transformations (and compilation)
+    in DaceProgram.__call_, this prevents other library nodes from being
+    expanded before the backward pass nodes.
+
+    This is clearly a hack and will be replaced with better way to achieve this.
+    """
+    def optimize(self):
+        expand_nodes(self.sdfg, lambda n: isinstance(n, BackwardPass))
```

Comment:
shouldn't it at least call `_current_opt` after this?

Reply:
Yes I will add that